src/client/game.cpp

```diff
@@ -555,6 +531,47 @@ class GameGlobalShaderConstantSetter : public IShaderConstantSetter
 		}
 	}
 
+public:
+	void onSettingsChange(const std::string &name)
+	{
+		if (name == "exposure_compensation")
+			m_user_exposure_compensation = g_settings->getFloat("exposure_compensation", -1.0f, 1.0f);
+	}
+
+	static void settingsCallback(const std::string &name, void *userdata)
+	{
+		reinterpret_cast<GameGlobalShaderConstantSetter*>(userdata)->onSettingsChange(name);
+	}
+
+	void setSky(Sky *sky) { m_sky = sky; }
+
+	GameGlobalShaderConstantSetter(Sky *sky, Client *client) :
+		m_sky(sky),
+		m_client(client)
+	{
+		for (auto &name : SETTING_CALLBACKS)
+			g_settings->registerChangedCallback(name, settingsCallback, this);
+
+		m_user_exposure_compensation = g_settings->getFloat("exposure_compensation", -1.0f, 1.0f);
+		m_bloom_enabled = g_settings->getBool("enable_bloom");
+		m_volumetric_light_enabled = g_settings->getBool("enable_volumetric_lighting") && m_bloom_enabled;
+	}
+
+	~GameGlobalShaderConstantSetter()
+	{
+		g_settings->deregisterAllChangedCallbacks(this);
+	}
+
+	void onSetConstants(video::IMaterialRendererServices *services) override
+	{
+		setMaterialConstants(services);
+		u64 frame_time = m_client->getEnv().getFrameTime();
+		if (frame_time != m_frame_time_prev) {
```

Comment:
Is this simple condition correct? I believe we'd need to check that the active (shader) program has stayed the same as well (or at least that we have uploaded the frame constant uniforms *for the given program* already): Uniforms are always tied (only) to the current active program.

Reply:
since the entire point of the PR hinges on this point I have the following thought: if we risk subtly breaking stuff by getting this condition wrong it may be better to drop the PR.

Reply:
With `/set time_speed 300000` and `/set fps_max 5` and intentionally breaking this condition with
```diff
diff --git a/src/client/game.cpp b/src/client/game.cpp
index 21f3a4401..f077e6357 100644
--- a/src/client/game.cpp
+++ b/src/client/game.cpp
@@ -568,7 +568,9 @@ class GameGlobalShaderConstantSetter : public IShaderConstantSetter
 		u64 frame_time = m_client->getEnv().getFrameTime();
 		if (frame_time != m_frame_time_prev) {
 			m_frame_time_prev = frame_time;
-			setFrameConstants(services);
+			if (rand() % 2 == 1) {
+				setFrameConstants(services);
+			}
 		}
 	}
 
```
, the nodes have different shading:
![randomised_setFrameConstants](https://github.com/user-attachments/assets/18c8101f-6b08-49dc-8e5e-4e08d28768c9)

Without intentionally breaking the condition, I haven't noticed problems yet.